keps/sig-node/20190423-expose-tty-probes.md

```diff
@@ -0,0 +1,123 @@
+
+## Summary
+
+Some readiness and liveness probes have exec commands that use `/bin/sh -i -c`, especially when deploying postgres and mysql. However, these probes fail with an `ioctl error` that happens because tty is set to false by default. Setting `-i` in thses exec command helps to pick the default binary paths and libraries from the start up scripts without having to specify the full paths in the command. Adding an optional boolean tty field to the probes gives the users the flexibility of turning n tty whenever using `/bin/sh -i -c` in their exec commands.
+
+## Motivation
+
+Make it easier for users that use liveness and readiness probes. This change will let the users set `/bin/sh -i -c` in their probes' exec commands without requiring them to specify the full path names of the commands they are running. They will no longer run into an `ioctl error` as the `tty` option will be exposed in the api.
+Users run into this problem especially when trying to deploy containers with databases such as postgres and mysql.
+
+### Goals
+
+Provide users with the ability to easily set an optional `tty` value when running liveness and readiness probes.
+
+## Proposal
+
+We propose adding a new boolean option called `Tty` in the CRI API for probes. This way users can easily configure it in their kube yaml files.
+
+```
+type Probe struct {
+	...
+	// Set tty for certain exec commands used in probes, especially when doing "/bin/sh -i -c".
+	// +optional
+	Tty bool
+}
+```
+
+The probes call an exec sync command when execing a command. So we will add `tty` as an option to the exec sync function that the probe calls.
+
+```
+ExecSync(containerID string, cmd []string, timeout time.Duration, tty bool) ([]byte, []byte, error)
+```
+
+The default value of the `tty` option will be false.
+
+Will also add tests for this to ensure it works and does not break any exisitng features.
+
+### Risks and Mitigations
+
+There are no serious risks with making `Tty` an optional arg for probes. Only users that run into the `ioctl error` will be using it.
+
+### Test Plan
+
+Will add an e2e test to the existing container probe tests.
+
+### Graduation Criteria
+
+stable:
+- No complaints about having the optional `Tty` arg for probes
+
+## Implementation History
+
+Major milestones in the life cycle of a KEP should be tracked in `Implementation History`.
+Major milestones might include
+
+- the `Summary` and `Motivation` sections being merged signaling SIG acceptance
+- the `Proposal` section being merged signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+
+## Alternatives [optional]
+
+- Setting tty to true by default in the exec sync function - this was rejected.
```

Comment:
Please link to discussion.

Reply:
Linked the PR where this discussion happened